### DISCUSSION TITLE
How to link created requirements with a logical system or with higher level requirements?

### DISCUSSION THREAD

**Discussion by dwdbmeagle** [dbinfrago/py-capellambse#482] [Q&A] [answered]
I am using capella 7 with the requirement viewpoint addon.
I have written a python script using py-capellambse which reads a table of requirements in a word doc and creates a set of matching requirements in capella in a relevant requirement_module using the requirement.create function:
`new_requirement = model.sa.requirement_modules[1].requirements.create(name=new_req_name,text=new_req_text)`
I would like to be able to do more, the input data includes the relevant system and the higher level associated requirement.
Is it possible to perform linking between existing requirements and logical components and other requirements using py-capellambse?
I have the relevant names and uuid's of the relevant higher lever requirements, logical components etc.
Which method or function should I use?

**Comment by Chang-Fab** [comment]
I'm also looking for a solution on this question.

**Comment by Wuestengecko** [comment, marked as answer]
Hi! You need to create `relations` on the new Requirement, like this:

````python
new_requirement = model.sa.requirement_modules[1].requirements.create(name=new_req_name,text=new_req_text)

related_element = model.by_uuid("...")
new_relation = new_requirement.relations.create(target=related_element)
````

Depending on the type of the target (i.e. whether it's a Requirement or another model element), this call will automatically pick between `InternalRelation` and `CapellaIncomingRelation`, so the same call works for both cases.

You can also assign relation types, either after creating the relation ...

````python
req_type = model.search("RequirementType", below=model.sa).by_name("Functional Requirement")

new_relation.type = req_type
````

... or directly during the `create()` call:

````python
new_relation = new_requirement.relations.create(target=related_element, type=req_type)
````

Let me know if this helps, or if something doesn't work as it should!

**Comment by Chang-Fab** [comment]
@Wuestengecko Thanks for your response!
Yes it works well, the relation has been created.
![image](https://github.com/user-attachments/assets/4b28973a-bc4b-4ed7-8ebf-8c8b63168c63)
How can I specify the relation type ? it is a **CapellaIncomingRelation** but I would like **CapellaOutcomingRelation** ?

**Reply by Wuestengecko** [reply]
CapellaOutgoingRelations live below the target element and link back to the Requirement. So they have to be created there as an extension instead, which is a bit more typing:

````python
related_element.extensions.create("CapellaOutgoingRelation", source=new_requirement, target=related_element)
````

(Note how you have to specify both `source` and `target` explicitly here, otherwise Capella wouldn't be able to recognize the relation properly.)

Afterwards, it will show up in the Requirement's `relations` too.

**Reply by Chang-Fab** [reply]
Thanks @Wuestengecko 
Unfortunately I get this error when I try you code
![image](https://github.com/user-attachments/assets/1f9a493b-0585-474f-906b-e961b60ddc4b)

**Reply by Wuestengecko** [reply]
`extensions` was introduced with capellambse v0.6.17, can you please make sure that you're on at least that version?

**Reply by Chang-Fab** [reply]
@Wuestengecko Thanks for your help and your reactivity.
Yeah, I used the version v0.6.15...
After an update of the version of `capellambse` package, it works well !!
You can resolve the question on my side because you've answered to all my questions related to the requirement allocation by Python !
![image](https://github.com/user-attachments/assets/f0149068-e6eb-49a6-8c55-9084ee3f0a18)

An aside point, do you have the doc to help me to update my code regarding the `FutureWarning`
![image](https://github.com/user-attachments/assets/be508332-9131-4e08-9ca7-c469f95632b4)

**Reply by Wuestengecko** [reply]
The warnings in that screenshot come from the context-diagrams extension, which hasn't been fully updated yet. I'll be working together with its primary maintainer, @ewuerger, to resolve those before removing the compatibility aliases in capellambse. So if you see such warnings and the file path says something like ``...\site-packages\capellambse_context_diagrams\...``, they're safe to ignore for now :)

**Reply by Chang-Fab** [reply]
@Wuestengecko Again, thanks for the support and the reactivity! Very appreciated !
